README.md
v1.4.0
=======
* Replaced Coins with Decimal Coins
* Alot of mod upgrades, mostly for 1.16
* Infinite Trading now require Collective

v1.3.0
========
* Added Hwyla (Here's what you're looking at)
* Added JEI (Just Enough Items)
* Added support for 1.12 and 1.14

v1.2.1
=========
* Updated Light Overlay 1.15-4.6.1 -> 1.15-5.4.0
* Updated Infinite Trading 1.15-1.0 -> 1.15-1.1
* Alpha release of 1.16: 7 of 16 mods included.

v1.2.0
=========
* Added redstone tools:
* Added Golden Hopper v1.1.1, a hopper with item-filter.
* Add Uppers 0.2.12, a up-side-down hopper, that can push items upwards.
* Add Workbench Factory v1.0.0, An hopper-enabled workbench.

v1.1.1
=========
* Version upgrades of Forge, Corpse & Bookshelf

v1.1.0
=========
* Added mod for help with lightning up rooms: [Light Overlay](https://www.curseforge.com/minecraft/mc-mods/light-overlay)
* Added library-mod: [Cloth Config API](https://www.curseforge.com/minecraft/mc-mods/cloth-config-forge)

v1.0.0
=========
* Added mod for RPG currency: [Coins](https://www.curseforge.com/minecraft/mc-mods/coins-je)
* Added library-mod [Bookshelf](https://www.curseforge.com/minecraft/mc-mods/bookshelf), required by Coins
* Added mod whit the ability to sit down: [Sit](https://www.curseforge.com/minecraft/mc-mods/sit)
* Added mod to prevent quest-villagers to get out of stock: [Infinite Trading](https://www.curseforge.com/minecraft/mc-mods/infinite-trading)
* Added mod to prevent players from removing spawners: [Unbreakable Spawner](https://www.curseforge.com/minecraft/mc-mods/unbreakable-spawner)

v0.0.3
=========
* Bugfix: Required = false, in the manifest.json, do wiered stuff, best to avoid it.

v0.0.2
=========
* Bugfix Minecraft-version and Forge-version didn't match
* Add missing dependecy [Placebo](https://www.curseforge.com/minecraft/mc-mods/placebo)

v0.0.1-r1
=========
* Fixed zipping issue, github releases puts all code in a subfolder, that format dosn't work with the twitch-app

v0.0.1
=========
* Added mod for not losing your equipment when you die: [Corpse](https://www.curseforge.com/minecraft/mc-mods/corpse)
* Added mod [journeymap](https://www.curseforge.com/minecraft/mc-mods/journeymap)
* Added mod for reduce lag from XP-particles: [Clumps](https://www.curseforge.com/minecraft/mc-mods/clumps)
* Added mod for reduce lag from Furnace: [FastFurnace](https://www.curseforge.com/minecraft/mc-mods/fastfurnace)
* Added mod for reduce lag from Workbench: [FastWorkbench](https://www.curseforge.com/minecraft/mc-mods/fastworkbench)



changelog.md
v1.3.0
========
* Added Hwyla (Here's what you're looking at)
* Added  JEI (Just Enough Items)
* Added support for 1.12 and 1.14

v1.2.1
=========
* Updated Light Overlay 1.15-4.6.1 -> 1.15-5.4.0
* Updated Infinite Trading 1.15-1.0 -> 1.15-1.1
* Alpha release of 1.16: 7 of 16 mods included.

v1.2.0
=========
* Added redstone tools:
* Added Golden Hopper v1.1.1, a hopper with item-filter.
* Add Uppers 0.2.12, a up-side-down hopper, that can push items upwards.
* Add Workbench Factory v1.0.0, An hopper-enabled workbench.

v1.1.1
=========
* Version upgrades of Forge, Corpse & Bookshelf

v1.1.0
=========
* Added mod for help with lightning up rooms: [Light Overlay](https://www.curseforge.com/minecraft/mc-mods/light-overlay)
* Added library-mod: [Cloth Config API](https://www.curseforge.com/minecraft/mc-mods/cloth-config-forge)

v1.0.0
=========
* Added mod for RPG currency: [Coins](https://www.curseforge.com/minecraft/mc-mods/coins-je)
* Added library-mod [Bookshelf](https://www.curseforge.com/minecraft/mc-mods/bookshelf), required by Coins
* Added mod whit the ability to sit down: [Sit](https://www.curseforge.com/minecraft/mc-mods/sit)
* Added mod to prevent quest-villagers to get out of stock: [Infinite Trading](https://www.curseforge.com/minecraft/mc-mods/infinite-trading)
* Added mod to prevent players from removing spawners: [Unbreakable Spawner](https://www.curseforge.com/minecraft/mc-mods/unbreakable-spawner)

v0.0.3
=========
* Bugfix: Required = false, in the manifest.json, do wiered stuff, best to avoid it.

v0.0.2
=========
* Bugfix Minecraft-version and Forge-version didn't match
* Add missing dependecy [Placebo](https://www.curseforge.com/minecraft/mc-mods/placebo)

v0.0.1-r1
=========
* Fixed zipping issue, github releases puts all code in a subfolder, that format dosn't work with the twitch-app

v0.0.1
=========
* Added mod for not losing your equipment when you die: [Corpse](https://www.curseforge.com/minecraft/mc-mods/corpse)
* Added mod [journeymap](https://www.curseforge.com/minecraft/mc-mods/journeymap)
* Added mod for reduce lag from XP-particles: [Clumps](https://www.curseforge.com/minecraft/mc-mods/clumps)
* Added mod for reduce lag from Furnace: [FastFurnace](https://www.curseforge.com/minecraft/mc-mods/fastfurnace)
* Added mod for reduce lag from Workbench: [FastWorkbench](https://www.curseforge.com/minecraft/mc-mods/fastworkbench)



updateCheck.php
<?php
declare(strict_types=1);

/**
 * @throws \JsonException
 * @throws \RuntimeException
 */
(static function () {
    $cache = [];
    $mods = \json_decode(file_get_contents(__DIR__ . '/src/mods.json'), false, 512, JSON_THROW_ON_ERROR);
    $types = [
        1 => 'Stable',
        2 => 'Beta',
        3 => 'Alpha',
    ];

    foreach ($mods as $modId => $modInfo) {
        $modId = +$modId;
        $lastFileId = max((array) $modInfo->files);
        sleep(1);
        $raw = file_get_contents("https://addons-ecs.forgesvc.net/api/v2/addon/{$modId}/files");
        if (!$raw) {
            echo 'ERROR fetching ', $modInfo->name, ' check ', $modInfo->url, PHP_EOL;
            continue;
        }
        try {
            $json = \json_decode($raw, false, 512, JSON_THROW_ON_ERROR);
        } catch (\JsonException $e) {
            echo 'ERROR parsing ', $modInfo->name, ' check ', $modInfo->url, PHP_EOL;
            echo $e->getMessage(), PHP_EOL;
            echo $raw;
            continue;
        }
        $cache[$modId] = $json;
        $newestFileId = max(array_column((array) $json, "id"));
        if ($newestFileId <= $lastFileId) {
            continue;
        }
        echo 'Updated ', $modInfo->name, ' check ', $modInfo->url, PHP_EOL;
        $updates = [];
        foreach ($json as $fileMeta) {
            $fileId = +$fileMeta->id;
            if($fileId <= $lastFileId) {
                continue;
            }
            $date = substr($fileMeta->fileDate, 0, 10);
            $gameVersions = implode(', ', $fileMeta->gameVersion);
            $type = $types[$fileMeta->releaseType] ?? $fileMeta->releaseType;
            $symbol = in_array('Forge', $fileMeta->gameVersion) ? '+' : (in_array('Fabric', $fileMeta->gameVersion) ? '-' : '*');
            $updates[$fileId] = " {$symbol} {$fileId} {$type} @ {$date}: {$fileMeta->displayName} for {$gameVersions}";
        }
        krsort($updates);
        foreach ($updates as $fielId => $udpate) {
            echo $udpate, PHP_EOL;
        }
        echo PHP_EOL;
    }
    file_put_contents(__DIR__ . '/build/cache.json', json_encode($cache, 64*7));
})();


release.php
<?php
declare(strict_types=1);

/**
 * @throws \JsonException
 * @throws \RuntimeException
 */
(static function () {
    $modVersion = '1.4.0';
    $mcVersions = [
        '1.12' => ['1.12.2', '14.23.5.2854'],
        '1.14' => ['1.14.4', '28.2.0'],
        '1.15' => ['1.15.2', '31.2.0'],
        '1.16' => ['1.16.1', '32.0.47'],
    ];
    $jsonOptions = JSON_THROW_ON_ERROR | JSON_PRETTY_PRINT | JSON_UNESCAPED_SLASHES | JSON_UNESCAPED_UNICODE;

    $mods = \json_decode(file_get_contents(__DIR__ . '/src/mods.json'), false, 512, JSON_THROW_ON_ERROR);
    $template = \json_decode(file_get_contents(__DIR__ . '/src/manifest.json'), false, 512, JSON_THROW_ON_ERROR);

    foreach ($mcVersions as $mcVersion => [$mcVersionPatch, $forgeVersion]) {
        $manifest = $template;
        $manifest->minecraft->version = $mcVersionPatch;
        $manifest->minecraft->modLoaders[0]->id = 'forge-' . $forgeVersion;
        $manifest->version = $modVersion;
        $manifest->files = [];

        foreach ($mods as $modId => $modInfo) {
            $fileId = $modInfo->files->$mcVersion ?? 0;
            if ($fileId) {
                $manifest->files[] = [
                    "projectID" => +$modId,
                    "fileID" => +$fileId,
                    "required" => true,
                ];
            }
        }

        $dir = __DIR__ . '/build/' . $mcVersion;
        if (!is_dir($dir) && !mkdir($dir, 0777, true) && !is_dir($dir)) {
            throw new \RuntimeException("Directory '{$dir}' was not created");
        }
        file_put_contents($dir . '/manifest.json', json_encode($manifest, $jsonOptions) . PHP_EOL);
        $zipFile = __DIR__ . "/releases/SundragonMods-{$mcVersion}-{$modVersion}.zip";
        $bashCommands = [
            'cd ' . escapeshellarg($dir),
            'zip ' . escapeshellarg($zipFile) . ' manifest.json',
        ];
        echo shell_exec(implode(' && ', $bashCommands)), PHP_EOL;
    }
})();
